Handle DeQuantStub() for QAT

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26627 Default observer and fake-quant for backends
* #26626 Moving average for statistics in observers
* #26625 Emulate weight and activation only quant with fake quant, numerics test
* #26624 Quantization aware training: Freeze batch norm support
* #26623 Per channel fake quant
* #26516 Per-channel baseline
* #26520 control of observer/fake-quant operations
* #26622 Default histogram observer
* #26519 Serialization and range reduction support for Fake Quant/Observer
* #26420 Fake quantization enhancements for QAT/PTQ support
* **#26518 Handle DeQuantStub() for QAT**

Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

raghuramank10000 committed:
Handle DeQuantStub() for QAT

Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -229,6 +232,8 @@ def forward(self, x):
     nn.LSTM: nnqd.LSTM,
 }
+# List of modules for which we do not need FakeQuant for activations
+DEFAULT_QAT_SKIP_MODULE_LIST = [DeQuantStub]
```
shall we do whitelist?

raghuramank100 replied:
This would need to be done for all the lists in quantize, so that the API is consistent. Shall we do this in a later PR?

jerryzh168 replied:
we can remove this

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

jerryzh168 commented:
discussed offline about removing skip_list

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -102,8 +102,9 @@ def add_observer(module):
         # observer and hook will be gone after we swap the module
-        module.add_module('observer', module.qconfig.activation())
-        module.register_forward_hook(_observer_forward_hook)
+        if module.training and type(module) not in set(DEFAULT_QAT_SKIP_MODULE_LIST) or module.training is False:
```
I think this is overcomplicated. just use`if not (module.training and type(module) == DeQuantStub)`

dzhulgakov replied:
this feels like a bit too magical to be so deep in the code. Why can't we just pass DeQuantStub as a skip list to propagate_qconfig? it already has a skip list

raghuramank100 replied:
Started with that design, with a list being passed to propagate_qconfig for the case of quantization aware training.  

Since we are going to change this to a whitelist (after 1.3), we can make this change later.  We need to land this soon, so lets settle on a preferred way soon. 

My recommendation is to have a list just so that the flow is clear to the end user and have it explicitly passed within prepare_qat calls (like the original version of the diff).

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

jerryzh168 commented:
@dzhulgakov the skip only happens when the module is in training mode

jerryzh168 commented:
also we plan to change it to whiltelist I think

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -102,8 +102,9 @@ def add_observer(module):
         # observer and hook will be gone after we swap the module
-        module.add_module('observer', module.qconfig.activation())
-        module.register_forward_hook(_observer_forward_hook)
+        if module.training and type(module) != DeQuantStub or module.training is False:
```
can you change this to `if not (module.training and type(module) == DeQuantStub)`

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/quantize.py:
```diff
@@ -11,8 +11,35 @@
+    r"""Dequantize stub module, before calibration, this is same as identity,
+    this will be swapped as `nnq.DeQuantize` in `convert`.
+    """
+    def __init__(self, qconfig=None):
```
then we don't need qconfig in DeQuantStub?

raghuramank100 replied:
Yes, removed it now

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Looks good (maybe check Jerry's comment)

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

raghuramank100 commented:
> Looks good (maybe check Jerry's comment)

Have addressed Jerry's comment, will land after tests pass.

jerryzh168 reviewed (approved):
LG Thanks

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Handle DeQuantStub() for QAT"


Skip Dequantize() modules for QAT alone. For fake quant insertion, DeQuantize() is a no-op and we should not be inserting fake-quant.

Differential Revision: [D17439333](https://our.internmc.facebook.com/intern/diff/D17439333/)

[ghstack-poisoned]
